Use a config file

max-torch opened the pull request:
In the context of app configuration, the app now reads from .json files if it's expected environment variables are unset. When running tests, the `pytest-env` dependency is now optional.

Closes #5

max-torch committed:
Add support for app deployment configuration options from a file

In the context of app configuration, the app now reads from .json files if it's expected environment variables are unset. When running tests, the `pytest-env` dependency is now optional.

max-torch committed:
Add missing quotation marks

soliton- commented:
I don't think it's useful to put the config files in a dot dir. We don't want to hide them.

Is it useful to have multiple config files? I could see a split between regular and security sensitive settings so that you can share/version/whatever one file and keep the other private but that's not the current split.

soliton- commented on .config/README.md:
````diff
@@ -0,0 +1,32 @@
+
+```json
+{
+    "url_base_pathname": "<your_url_base_pathname>",  // The pathname must start and end with "/". This field is optional and defaults to "/" when undeclared.
````
The comment makes the JSON invalid. I suggest to move the comment elsewhere or use a format that supports comments.

soliton- commented on .config/README.md:
```diff
@@ -0,0 +1,32 @@
+
+You can also use environment variables in place of `db_config.json`, as specified [here](../README.md#using-environment-variables-for-database-credentials-optional).
+
+__*These files are not tracked in Git because they can contain sensitive information.__
```
IMO the more important part is that the content will be different for everyone. Doesn't make much sense to track user config here. (Users could track it for themselves in a different repo though.)

soliton- commented on app.py:
```diff
@@ -256,7 +272,9 @@ def update_total_games_value(start_date, end_date):
     mariadb_connection = connect_to_mariadb()
     cursor = mariadb_connection.cursor()
-    target_table = "wesnothd_game_info"
+    with open(".config/config.json", "r") as f:
```
I would open the config file once and not multiple times at different places in the code.

soliton- commented on app.py:
```diff
@@ -11,7 +12,10 @@
+    Connects to a MariaDB database using environment variables or a .json file for authentication.
+
+    The function first tries to load configuration from environment variables.
+    If any configuration is missing, it tries to load from a .json file.
```
That is unexpected. The common behavior is to first consider defaults (not everything can/should have defaults of course, here the DB port could default to 3306 and DB host to localhost) then read the config file then let env variables override settings and then let command line parameters override settings last.

So defaults get overridden by config file gets overridden by env variables gets overridden by command line parameters.

max-torch committed:
Use a memoized function to get the table name

max-torch committed:
Change priority of configuration options interface

max-torch committed:
Necessary refactor to get Flask-Caching to work

The code had to be refactored in order to get Flask-Caching to work for both production and development contexts.

For Flask-Caching to work, the cache initialization and the `server` object instantiation seems to need to be in the same file, and both seem to need to be unnested top-level code as well.

max-torch committed:
Refactor configuration files

max-torch commented:
Please note that the app deployment instructions have now changed very slightly.
Instead of `gunicorn --bind :$PORT main:server` you will now have to use `gunicorn --bind :$PORT app:server`.

soliton- commented on app.py:
```diff
@@ -1,17 +1,65 @@
+    logging.debug(
+        "get_target_table called"
+    )  # This only gets logged the first time the function is called and proves that memoization is functioning.
+    with open("config.json", "r") as f:
```
This opens the config file every time this function is called?

max-torch replied:
only the first time the function is called because of `@cache.memoize()`
Confirmed by the debug log statement.
er, perhaps the log message should be rewritten to be more clear?
Log instead that the file was read? I realize it seems inaccurate, the log, because the function does get called again after the first time but just not executed.

I'll change the wording.

max-torch replied:
Please now see https://github.com/wesnoth/wesnoth-multiplayer-data-dashboard/pull/10/commits/9bbf9f8d2652b918d58298d1b6aac3f4ff41c97b, the statement has been changed.

soliton- replied:
This seems pretty confusing. I would normally expect loading the config file once at program start into some internal config object and using that throughout the app via a function or whatever.

Anyway if you like it like this that's fine.

max-torch committed:
Update README.md

small correction

max-torch committed:
Update app.py

More accurate log statement

soliton- commented on README.md:
````diff
@@ -144,7 +173,7 @@ To serve this app to end users, it is recommended to use a WSGI server such as [
 
 ```bash
-gunicorn main:server
+gunicorn --bind :$PORT app:server
````
I think it'd be clearer to use `--bind 0.0.0.0:$PORT` since that appears to be what leaving out the host part does.

max-torch committed:
Update README.md

more explicit example

max-torch commented:
PR looks to be in a good place now. I will merge to main in 2 days.
